phoenix-core/src/main/java/org/apache/phoenix/schema/stats/GuidePostsInfoBuilder.java

```diff
@@ -17,124 +17,112 @@
+     * Guide posts organized in chunks
      */
-    private int maxLength;
-    private DataOutputStream output;
-    private TrustedByteArrayOutputStream stream;
-    private List<Long> rowCounts = new ArrayList<Long>();
-    private List<Long> byteCounts = new ArrayList<Long>();
-    private List<Long> guidePostsTimestamps = new ArrayList<Long>();
-
-    public boolean isEmpty() {
-        return rowCounts.size() == 0;
-    }
-    
-    public List<Long> getRowCounts() {
-        return rowCounts;
-    }
+    private List<GuidePostChunk> guidePostChunks;
 
-    public List<Long> getByteCounts() {
-        return byteCounts;
-    }
+    /**
+     * Builder for the guide post chunk
+     */
+    private GuidePostChunkBuilder chunkBuilder;
 
-    public List<Long> getGuidePostsTimestamps() {
-        return guidePostsTimestamps;
-    }
+    /**
+     * The index of the guide post chunk in the chunk array
+     */
+    private int guidePostChunkIndex;
 
-    public int getMaxLength() {
-        return maxLength;
-    }
-    public GuidePostsInfoBuilder(){
-        this.stream = new TrustedByteArrayOutputStream(1);
-        this.output = new DataOutputStream(stream);
-        this.encoder=new PrefixByteEncoder();
-        lastRow = new ImmutableBytesWritable(ByteUtil.EMPTY_BYTE_ARRAY);
+    /**
+     * Use to track the lower bound of generated guide post chunks.
+     * It's always exclusive.
+     */
+    private ImmutableBytesWritable lastRow;
+
+    /**
+     * The total count of guide posts collected
+     */
+    private int guidePostsCount;
+
+    public GuidePostsInfoBuilder() {
+        this(1);
     }
 
-    public boolean addGuidePostOnCollection(ImmutableBytesWritable row, long byteCount,
-            long rowCount) {
-        /*
-         * When collecting guideposts, we don't care about the time at which guide post is being
-         * created/updated at. So passing it as 0 here. The update/create timestamp is important
-         * when we are reading guideposts out of the SYSTEM.STATS table.
-         */
-        return trackGuidePost(row, byteCount, rowCount, 0);
+    public GuidePostsInfoBuilder(int targetedChunkSize){
+        this.targetedChunkSize = targetedChunkSize;
+        this.guidePostChunks = Lists.newArrayListWithExpectedSize(1);
+        this.guidePostChunkIndex = 0;
+        this.lastRow = new ImmutableBytesWritable(KeyRange.UNBOUND);
+        this.guidePostsCount = 0;
     }
 
     /**
      * Track a new guide post
      * @param row number of rows in the guidepost
      * @param byteCount number of bytes in the guidepost
      * @param updateTimestamp time at which guidepost was created/updated.
-     * @throws IOException
      */
-    public boolean trackGuidePost(ImmutableBytesWritable row, long byteCount, long rowCount,
-            long updateTimestamp) {
+    public boolean trackGuidePost(ImmutableBytesWritable row, long byteCount, long rowCount, long updateTimestamp) {
```

Comment:
This isn't really following the builder pattern.  consider rename or handling through exceptions or changing the caller.

Reply:
This intend not to use builder pattern here. No need to use builder pattern.